### PR TITLE
[MIRROR] [MDB IGNORE] dir sanity, primarily on WALLITEMs [MDB IGNORE]

### PR DESCRIPTION
**Pull request by SkyratBot**
Original PR: https://github.com/tgstation/tgstation/pull/62601
--------------------

## About The Pull Request

Wall items *mostly* use the direction from the floor to the wall in the named mapping helper. Wall items *mostly* use the direction from the wall to the floor for the internal `dir` variable.

This leads to a headache when it comes to working out what conflicts with what, and what needs placing where.

Wall frames provided a member, `inverse`, which specified whether or not to invert the direction of the item when looking for conflicts. It was also used to specify whether to look for conflicts outside of the wall (cameras and lights appear external to the wall) or inside the wall (most wall items). This flag was set for Intercoms, APCs, and Lights. Since APCs and Lights expect a floor-to-wall direction, and Intercoms expect a wall-to-floor direction, this means that APCs and Lights were getting the correct direction, and Intercoms were getting the wrong direction.

Some implications of this setup were:

- You could build an APC on top of another wall item, provided there was nothing external attached to the wall and the area didn't have an APC.
- You could stack Intercoms indefinitely on top of the same wall, provided you weren't in a one-tile wide corridor with something on the opposite wall.

Or both! Here's twenty Intercoms placed on the wall, and a freshly placed APC frame after placing all Intercoms and deconstructing the old APC:

endless-stack-of-intercoms

Not everything used this `inverse` variable to adjust to the correct direction. For example, `/obj/machinery/defibrillator_mount` just used a negative `pixel_offset` to be visually placed in the correct direction, even though the internal direction was wrong, and never set! This also let you stack an indefinite number of defib mounts on the same wall, provided it wasn't a northern wall... except you could do this to northern walls too, since defibs weren't considered a wall item for the purposes of checking collisions at all!

Ultimately, every constructable interior wall item either used this `inverse` variable to adjust to the correct placement, set a negative `pixel_offset` variable to have its offset adjusted to the correct placement, or overrode `New` or `Initialize` to run its own checks and assignment to `pixel_x` and `pixel_y`!

<details>
<summary>Inventory: Table of various paths, related paths, and the adjustments they used</summary>

Construction Path|Path|Related Paths|Name|global_lists.dm|global_lists.dm inverted|Frame inverted|Pixel offset inverted|Name direction|Dir direction
--- | --- | --- | --- | --- | --- | --- | --- | --- | ---
`/obj/item/wallframe/intercom`|`/obj/item/radio/intercom`||Intercom|Interior|yes|yes||F2W|W2F
[unbuildable]|`/obj/item/storage/secure/safe` ||Secure Safe|Interior||||F2W|W2F
[unbuildable]|`/obj/item/storage/secure/safe/caps_spare`||Spare ID Safe|Interior||||[pixel offsets only]|Always South
`/obj/item/wallframe/airalarm`|`/obj/machinery/airalarm` ||Air Alarm|Interior|||yes|F2W|W2F
`/obj/item/wallframe/bounty_board`|`/obj/machinery/bounty_board` ||Bounty Board|Interior|||yes|F2W|W2F
`/obj/item/wallframe/button`|`/obj/machinery/button`||Button|Interior|||yes|F2W|W2F
[unbuildable]|`/obj/machinery/computer/security/telescreen`||Telescreen|Interior||||[unnamed]|W2F
[unbuildable]|`/obj/machinery/computer/security/telescreen/entertainment`||Entertainment Monitor|Interior||||F2W|W2F
[unbuildable]|`/obj/machinery/door_timer`||Door Timer|Interior||||[pixel offsets only]|Always South
[unbuildable]|`/obj/machinery/embedded_controller/radio/simple_vent_controller`||Simple Vent Controller|Interior||||[unused]|[unused]
[unbuildable]|`/obj/machinery/embedded_controller/radio/airlock_controller`||Incinerator Access Console|None||||[pixel offsets only]|Always South
`/obj/item/wallframe/firealarm`|`/obj/machinery/firealarm`||Fire Alarm|Interior||yes||F2W|W2F
`/obj/item/wallframe/flasher`|`/obj/machinery/flasher`||Flasher|Interior|||yes|F2W|W2F [inconsistent]
[unbuildable]|`/obj/machinery/keycard_auth`||Keycard Authenticator|Interior||||F2W|W2F
[unbuildable]|`/obj/machinery/light_switch`||Light Switch|Interior||||F2W|W2F
`/obj/item/wallframe/newscaster`|`/obj/machinery/newscaster`||Newscaster|Interior||||F2W|Always South
`/obj/item/wallframe/apc`|`/obj/machinery/power/apc`||APC|Interior|yes|yes||F2W|F2W
[unbuildable]|`/obj/machinery/requests_console`||Requests Console|Interior||||F2W|W2F
`/obj/item/wallframe/status_display`|`/obj/machinery/status_display`||Status Display|Interior|||yes|F2W|Always South
`/obj/item/wallframe/extinguisher_cabinet`|`/obj/structure/extinguisher_cabinet`||Extinguisher Cabinet|Interior|||yes|F2W|W2F
[unbuildable]|`/obj/structure/fireaxecabinet`||Fire Axe Cabinet|Interior||||F2W|W2F
[unbuildable]|`/obj/structure/mirror`||Mirror|Interior||||F2W|W2F
[unbuildable]|`/obj/structure/noticeboard`||Notice Board|Interior||||F2W|W2F
[unbuildable]|`/obj/structure/reagent_dispensers/peppertank`||Pepper Spray Dispenser|Interior||||F2W|W2F
[unbuildable]|`/obj/structure/sign`||Sign|Interior||||[on wall]|Always South, even when wallitem
`/obj/item/wallframe/picture`|`/obj/structure/sign/picture_frame`||Sign/Picture Frame|Interior|||yes|[pixel offsets only]|Always South
`/obj/item/wallframe/camera`|`/obj/machinery/camera`|`/obj/structure/camera_assembly`|Camera|Exterior||||[unnamed]|W2F
`/obj/item/wallframe/light_fixture`|`/obj/machinery/light`|`/obj/structure/light_construct`|Light|Exterior|yes|yes||F2W|F2W
`/obj/item/wallframe/defib_mount`|`/obj/machinery/defibrillator_mount`||Defibrillator|None|||yes|F2W|W2F
`/obj/item/wallframe/bluespace_vendor_mount`|`/obj/machinery/bluespace_vendor`||Bluespace Gas Vendor|None|||yes|F2W|W2F
`/obj/item/wallframe/turret_control`|`/obj/machinery/turretid`||Turret Control|None||||[pixel offsets only]|Always South
`/obj/item/wallframe/painting`|`/obj/structure/sign/painting`||Painting|None||||[pixel offsets only]|Always South except on the lavaland syndie base, where it’s F2W
[unbuildable]|`/obj/machinery/ticket_machine`||Ticket Machine|None||||[on wall]|
</details>

Unfortunately, untangling `/obj/structure/sign` is going to be another major headache, and this has already exploded in scope enough already, so we can't get rid of the `get_turf_pixel` call just yet. This also doesn't fix problems with the special 2x1 `/obj/structure/sign/barsign`.

Some non-wall items have been made to use the new `MAPPING_DIRECTIONAL_HELPERS` as part of the directional cleanup.

---

tl;dr: All wall mounted items and some directional objects now use the same direction that they were labelled as. More consistent directional types everywhere.

## Why It's Good For The Game

fml

## Changelog

:cl:
refactor: Wall mounted and directional objects have undergone major internal simplification. Please report anything unusual!
fix: You can no longer stack an indefinite amount of Intercoms on the same wall.
fix: Defibrillator Mounts, Bluespace Gas Vendors, Turret Controlers, and Ticket Machines are now considered wall items.
fix: Wall mounted items on top of the wall now consistently check against other items on top of the wall, and items coming out of the wall now consistently check against other items coming out of the wall.
fix: The various directional pixel offsets within an APC, Fire Extinguisher Cabinet, Intercom, or Newscaster have been made consistent with each other.
fix: The pixel offsets of Intercoms, Fire Alarms, Fire Extinguisher Cabinets, Flashers, and Newscasters have been made consistent between roundstart and constructed instances.
fix: Constructed Turret Controls will no longer oddly overhang the wall they were placed on.
qol: Defibrillator mounts now better indicate which side of the wall they are on.
fix: Some instances where there were multiple identical lights on the same tile have been fixed to only have one.
/:cl:

### CODE REVIEW

**Comment by GoldenAlpharex** [comment]
esainane The Unspaghettifier.
